---
title: "R introduction and exploratory data analysis with GEO"
author: ""
date: ''
output:
  html_document:
    df_print: paged
---

```{r global options, include = FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

## Gene Expression Omnibus

We will dive straight into loading a gene expression data (made with 2 channel microarrays). GEO contains all sorts of high-throughput experiments, although microarray and RNA-seq are the most popular. 

> The Gene Expression Omnibus (GEO) project was initiated at NCBI in 1999 in response to the growing demand for a public repository for data generated from high-throughput microarray experiments. GEO has a flexible and open design that allows the submission, storage, and retrieval of many types of data sets, such as those from high-throughput gene expression, genomic hybridization, and antibody array experiments. 

Please explore and search your favorite terms on https://www.ncbi.nlm.nih.gov/geo/
As you follow along, it could be great for you to explore your own dataset (besides the one I use as an example).

While you can manually download and load the text file into R, it is easier to use the API/R package called GEOquery to do this automatically. To download the dataset on GEO directly into R, install these packages, and their dependencies:

```{r setup, results=FALSE}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("Biobase")
# BiocManager::install("GEOquery")

library(Biobase)
library(GEOquery)
```

We are interested in the following dataset called GDS39 coming from the classic experiments on cell cycle regulation in yeast.

> Spellman PT, Sherlock G, Zhang MQ, Iyer VR et al. Comprehensive identification of cell cycle-regulated genes of the yeast Saccharomyces cerevisiae by microarray hybridization. Mol Biol Cell 1998 Dec;9(12):3273-97. PMID: 9843569
Explore GEO and download directly from R:

The following code is used to download and look at meta data:

```{r geo1}
#Loading a GDS file with GEOquery
dat <- getGEO('GDS39', destdir=".")

# lets look at the class and the mode of this variable, dat
class(dat)
mode(dat)

# look at ATTRIBUTES of meta data, such that you know what information are available
attributes(Meta(dat))

# save and selectively look at meta data
metadat <- Meta(dat)
metadat$description
```

## Accessing the gene expression data

We can access the gene expression data using Table. See how we look at column names and row names to ensure what observations and variables are available. 

```{r geo2}
# we have a relatively large dataset, that will likely difficult to see at once
dim(Table(dat))

# then, look at a small portion of gene expression 
# note that the first 2 columns are identifiers for genes
Table(dat)[1:10,1:5]

# you can also access the whole row by not specifying the column number
Table(dat)[1,]

# column names are sample IDs 
colnames(Table(dat))

# you could selectively call on a column by its name 
# and also specify which rows to view (e.g., the first 10 rows here)
Table(dat)[1:10,"GSM910"]

```

There are numerous ways to slice and group the large matrix. Two R packages that are handy for this task is ``data.table``. For a starter, it'll nicely format the display of a large data. We will use it in the future! 

```{r datatable}
# install.packages("data.table")
library(data.table)
geneexp.dt <- as.data.table(Table(dat))

# note that data.table is based on data.frame
# all functionalities of data.frame (such as slicing) are available to data.table
class(geneexp.dt)

# we can also programmatically call on particular genes "FHL1"
geneexp.dt[IDENTIFIER == "FHL1",]

# call on "FHL1", at particular time points
geneexp.dt[IDENTIFIER == "FHL1", c("GSM940", "GSM942", "GSM910")]
```

## Data visualization

We look at one time point (one column) of this dataset. Particularly, we want to see the summary statistics. Then, a histogram, a density plot, and a boxplot are made. Please try on other columns, and multiple columns:

```{r vis1}
## look at one column (first tp) GSM940
GSM940 <- as.numeric(Table(dat)[,3])
summary(GSM940)

hist(GSM940)
hist(GSM940, breaks=100, xlim=c(-5,5), main="Histogram of 1st Time Point in the Yeast Study")

## density estimation and visualization
density.est = density(GSM940, kernel="gaussian", na.rm=T)
plot(density.est, xlim=c(-5,5))

## look at one column (first tp) GSM940
par(mfrow=c(1,2))
boxplot(GSM940, ylim=c(-5,5))
```

## Tidy data and ggplot2

We will re-structure this gene expression data into a tidy data. For further and well-written summary of tidy data, please see http://vita.had.co.nz/papers/tidy-data.html
Using this tidy data, ggplot2 can be used to create some beautiful visualizations fast. You need to install necessary R packages, which are packaged into tidyverse: https://tidyverse.tidyverse.org/:

```{r ggplot2}
# install.packages("tidyverse")
# more info and troubleshooting available https://tidyverse.tidyverse.org/
library(tidyverse)

geneexp <- Table(dat)
geneexp.tidy <- gather(geneexp, key="Samples", value="GeneExp", -c(1,2))

# let's make a box plot for all samples combined
ggplot(geneexp.tidy, aes(x=0,y=GeneExp)) + geom_boxplot() + ylim(-5,5)

# we will make a box plot for each samples (e.g., time points).
# note the use of "facet_grid" 
ggplot(geneexp.tidy, aes(x=0,y=GeneExp)) + geom_boxplot() + ylim(-5,5) + facet_grid(~Samples)

# how about a violin plot for each samples (e.g., time points)
# a violin plot:  https://en.wikipedia.org/wiki/Violin_plot
ggplot(geneexp.tidy, aes(x=0,y=GeneExp)) + geom_violin() + ylim(-5,5) + facet_grid(~Samples)

# how about a series of histogram?
ggplot(geneexp.tidy, aes(GeneExp)) + geom_histogram(bins=30) + xlim(-5,5)  + facet_grid(~Samples)

# save ggplot, especially if your monitor is too small to see many facets
# use ggsave()
```

The tidyverse is an extensive network of packages that use common data structures and lingos to make exploration and visualization simple. It is a learning process that will benefit you in a long run. There are a lot of intuitive and helpful documentations online. 
See the cheatsheet on data import and structure manipulation: https://github.com/rstudio/cheatsheets/blob/master/data-import.pdf
Or the datacamp has an interactive R course on tidyverse: https://www.datacamp.com/courses/introduction-to-the-tidyverse

## Heatmaps using gplots

A heatmap represents the individual values in a matrix as colors. For us, it usually allow the large table of numeric values to be visualized. Often, clustering or other organization of rows and columns help reveal patterns. Look and clean up any missing values. For this, we will simply remove any row that has missing values.

```{r missing}
dat.geneexp <- Table(dat)[,-c(1,2)]
dat.geneexp <- as.data.frame(lapply(dat.geneexp, as.numeric))
dat.geneexp <- apply(dat.geneexp,2,as.numeric)

# see if it has any missing values
sum(is.na(dat.geneexp))

# find rows with missing values
rows_missing <- apply(dat.geneexp, 1, function(x) any(is.na(x)))
dat.geneexp.complete <- dat.geneexp[!rows_missing,]
```

Let's create a heatmap with ``gplots``.

```{r heatmap}
library(RColorBrewer)
library(gplots)
my_palette <- colorRampPalette(c("blue", "white", "yellow"))(n = 299)

heatmap.2(dat.geneexp.complete,
          main = "Heatmap", # heat map title
          notecol="black",      # change font color of cell labels to black
          density.info="none",  # turns off density plot inside color legend
          trace="none",         # turns off trace lines inside the heat map
          margins =c(1,1),     # widens margins around plot
          col=my_palette,       # use on color palette defined earlier 
          dendrogram="none",     # only draw a row dendrogram
          scale = "row",
          Colv=FALSE)
```

## Homework

Please try creating a heatmap using ggplot2.
```{r homework}
# find rows with missing values
sum(is.na(geneexp.tidy))
rows_missing <- apply(geneexp.tidy, 1, function(x) any(is.na(x)))
geneexp.tidy.complete <- geneexp.tidy[!rows_missing,]
is.data.frame(geneexp.tidy.complete)
sum(is.na(geneexp.tidy.complete))

geneexp.tidy.complete$GeneExp <- scale(as.numeric(geneexp.tidy.complete$GeneExp))

ggplot(geneexp.tidy.complete, aes(x = Samples, y = IDENTIFIER, fill = GeneExp))+
  geom_tile() +
  xlab("Genes") +
  ylab("Samples") +
  theme(axis.text.y = element_blank(),
    axis.text.x = element_text(angle = 45, vjust = 0.5)) + # Rotate the x-axis labels
  scale_fill_gradient2(low = "blue", mid = "white", high = "yellow", midpoint = 0, limits=c(-3,3))
ggsave("lukasik_hw1_heatmap.png")
```